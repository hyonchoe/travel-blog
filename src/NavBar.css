.userMenus {
    float: right;
}

.userInfo {
    pointer-events: none;
    font-style: italic;
}