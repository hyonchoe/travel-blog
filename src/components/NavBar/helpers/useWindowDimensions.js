import { useState, useEffect } from 'react';

const useWindowDimensions = () => {
  const [windowWidth, setWindowWidth] = useState(getWindowWidth());

  useEffect(() => {
    const handleResize = () => {
        setWindowWidth(getWindowWidth())
    }

    window.addEventListener('resize', handleResize)
    return () => window.removeEventListener('resize', handleResize)
  }, [])

  return { windowWidth }
}

const getWindowWidth = () => {
    return window.innerWidth
}

export default useWindowDimensions