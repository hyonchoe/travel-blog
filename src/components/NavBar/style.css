.userMenus {
    float: right;
}

.userInfo {
    margin-right: 10px;
}