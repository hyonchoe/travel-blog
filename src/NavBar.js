import React from 'react'
import { Link, useLocation } from 'react-router-dom'
import { Menu } from 'antd'
import { useAuth0 } from '@auth0/auth0-react'
import './NavBar.css'

const NavBar = () =>{
    const location = useLocation()
    const { isAuthenticated, user, loginWithRedirect, logout } = useAuth0()

    const handleClick = (e) => {
        if (e.key === 'login' || e.key === 'logout'){
            (isAuthenticated) ? logout() : loginWithRedirect()
        }
    }

    return (
        <Menu 
            theme="dark"
            mode="horizontal"
            selectedKeys={[location.pathname]}
            onClick={handleClick} >
            <Menu.Item key="/">
                <Link to="/">Travelers' Feed</Link>
            </Menu.Item>
            {isAuthenticated && (
            <Menu.Item key="/myTrips">
                <Link to="/myTrips">My Trips</Link>
            </Menu.Item>                
            )}
            {isAuthenticated && ( 
            <Menu.Item key="/addTrip">
                <Link to="/addTrip"> Add/Edit Trip </Link>
            </Menu.Item>
            )}

            {!isAuthenticated && (
            <Menu.Item key='login' className="userMenus"> Log in </Menu.Item>
            )}
            {isAuthenticated && (
            <Menu.Item key='logout' className="userMenus"> Log out </Menu.Item>
            )}
            {isAuthenticated && (
            <Menu.Item key='userName' className="userMenus userInfo" >Hello, {(user.name) ? user.name : user.email}</Menu.Item>
            )}
        </Menu> 
    )   
}

export default NavBar