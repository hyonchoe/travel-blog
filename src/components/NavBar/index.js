import React from 'react'
import { Link, useLocation } from 'react-router-dom'
import { useAuth0 } from '@auth0/auth0-react'
import { Menu, Button } from 'antd'
import { MenuOutlined } from '@ant-design/icons'
import useWindowDimensions from './useWindowDimensions'
import './style.css'


const breakingPointMd = 768

const NavBar = () =>{
    const location = useLocation()
    const { isAuthenticated, user, logout, loginWithPopup } = useAuth0()
    const { windowWidth } = useWindowDimensions()
    const collapsed = useCollapsedMenu(windowWidth)

    const UserMenuItem = () => {
        if (collapsed){
            return (
            <Menu.Item>
                {!isAuthenticated && (
                    <Button onClick={()=>loginWithPopup()}>Log in</Button>
                )}
                {isAuthenticated && (
                    <Button onClick={()=>logout()}>Log out</Button>
                )}
            </Menu.Item>               
            )
        }
       
        return (
            <Menu.Item
                className="userMenus" >
                {!isAuthenticated && (
                    <Button onClick={()=>loginWithPopup()}>Log in</Button>
                )}
                {isAuthenticated &&  (
                UserNameSpan()
                )}
                {isAuthenticated && (
                    <Button onClick={()=>logout()}>Log out</Button>
                )}
            </Menu.Item>
        )
    }
    const TravelerFeedMenuItem = () => {
        return (
            <Menu.Item key="/">
                <Link to="/">Travelers' Feed</Link>
            </Menu.Item>
        )
    }
    const MyTripsMenuItem = () => {
        if (isAuthenticated){
            return (
                <Menu.Item key="/myTrips">
                    <Link to="/myTrips">My Trips</Link>
                </Menu.Item>
            )
        }
        
        return null
    }
    const AddTripsMenuItem = () => {
        if (isAuthenticated){
            return (
                <Menu.Item key="/addTrip">
                    <Link to="/addTrip"> Add/Edit Trip </Link>
                </Menu.Item>
            )
        }

        return null
    }
    const UserNameMenuItem = () => {
        if(isAuthenticated){
            return (
                <Menu.Item
                    className="userMenus" >
                    {UserNameSpan()}
                </Menu.Item>
            )
        }

        return null
    }
    const UserNameSpan = () => {
        return (
            <span className="userInfo">Hello, {(user.name) ? user.name : user.email}</span>
        )
    }

    return (
        <Menu
            theme="dark"
            mode="horizontal"
            selectedKeys={[location.pathname]} >
            { collapsed &&
            <Menu.SubMenu
                icon={<MenuOutlined />}
                className="collapsedMenu" >
                {TravelerFeedMenuItem()}
                {MyTripsMenuItem()}
                {AddTripsMenuItem()}
                {UserMenuItem()}
            </Menu.SubMenu>
            }
            { collapsed &&
            UserNameMenuItem()
            }
            
            { !collapsed &&
            TravelerFeedMenuItem()
            }
            { !collapsed && 
            MyTripsMenuItem()
            }
            { !collapsed && 
            AddTripsMenuItem()
            }
            { !collapsed &&
            UserMenuItem()
            }

        </Menu> 
    )   
}

const useCollapsedMenu = (windowWidth) => {
    return windowWidth < breakingPointMd
}

export default NavBar