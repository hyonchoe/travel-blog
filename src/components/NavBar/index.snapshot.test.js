import React from 'react'
import { shallow } from 'enzyme'
import * as hooks from '@auth0/auth0-react'
import * as windowHooks from './useWindowDimensions'
import NavBar from './'

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useLocation: () => ({
        pathname: 'dummypathname'
    })
}))

describe('NavBar', () => {
    describe('for full menu', () =>{
        beforeEach(() => {
            jest.spyOn(windowHooks, 'default').mockImplementation(() => ({
                windowWidth: 1000
            }))
        })

        it('matches snapshot authenticated', () => {
            jest.spyOn(hooks, 'useAuth0').mockImplementation(() => ({
                isAuthenticated: true,
                user: { name: 'dummyname', email: 'dummyemail', sub: 'dummysub' },
                logout: jest.fn(),
                loginWithPopup: jest.fn()
            }))
            const wrapper = shallow(
                <NavBar />
            )
            expect(wrapper).toMatchSnapshot()
        })
    
        it('matches snapshot not authenticated', () => {
            jest.spyOn(hooks, 'useAuth0').mockImplementation(() => ({
                isAuthenticated: false,
                user: null,
                logout: jest.fn(),
                loginWithPopup: jest.fn()
            }))
            const wrapper = shallow(
                <NavBar />
            )
            expect(wrapper).toMatchSnapshot()
        })
    })
    
    describe('for collapsed menu', () => {
        beforeEach(() => {
            jest.spyOn(windowHooks, 'default').mockImplementation(() => ({
                windowWidth: 400
            }))
        })

        it('matches snapshot authenticated', () => {
            jest.spyOn(hooks, 'useAuth0').mockImplementation(() => ({
                isAuthenticated: true,
                user: { name: 'dummyname', email: 'dummyemail', sub: 'dummysub' },
                logout: jest.fn(),
                loginWithPopup: jest.fn()
            }))
            
            const wrapper = shallow(
                <NavBar />
            )
            expect(wrapper).toMatchSnapshot()
        })
    
        it('matches snapshot not authenticated', () => {
            jest.spyOn(hooks, 'useAuth0').mockImplementation(() => ({
                isAuthenticated: false,
                user: null,
                logout: jest.fn(),
                loginWithPopup: jest.fn()
            }))
            const wrapper = shallow(
                <NavBar />
            )
            expect(wrapper).toMatchSnapshot()
        })
    })
})