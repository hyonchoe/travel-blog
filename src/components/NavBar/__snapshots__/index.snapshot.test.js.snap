// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar matches snapshot authenticated 1`] = `
<Menu
  mode="horizontal"
  selectedKeys={
    Array [
      "dummypathname",
    ]
  }
  theme="dark"
>
  <MenuItem
    key="/"
  >
    <Link
      to="/"
    >
      Travelers' Feed
    </Link>
  </MenuItem>
  <MenuItem
    key="/myTrips"
  >
    <Link
      to="/myTrips"
    >
      My Trips
    </Link>
  </MenuItem>
  <MenuItem
    key="/addTrip"
  >
    <Link
      to="/addTrip"
    >
       Add/Edit Trip 
    </Link>
  </MenuItem>
  <MenuItem
    className="userMenus"
  >
    <span
      className="userInfo"
    >
      Hello, 
      dummyname
    </span>
    <Button
      block={false}
      ghost={false}
      htmlType="button"
      loading={false}
      onClick={[Function]}
    >
      Log out
    </Button>
  </MenuItem>
</Menu>
`;

exports[`NavBar matches snapshot not authenticated 1`] = `
<Menu
  mode="horizontal"
  selectedKeys={
    Array [
      "dummypathname",
    ]
  }
  theme="dark"
>
  <MenuItem
    key="/"
  >
    <Link
      to="/"
    >
      Travelers' Feed
    </Link>
  </MenuItem>
  <MenuItem
    className="userMenus"
  >
    <Button
      block={false}
      ghost={false}
      htmlType="button"
      loading={false}
      onClick={[Function]}
    >
      Log in
    </Button>
  </MenuItem>
</Menu>
`;
